pkg/apis/aws/validation/filter.go

```diff
@@ -0,0 +1,94 @@
+	// ZoneNameRegex matches e.g. us-east-1a
+	ZoneNameRegex = `^[a-z0-9-]+$`
+	// TagKeyRegex matches Letters (a–z, A–Z), numbers (0–9), spaces, and the following symbols: + - = . _ : / @
+	TagKeyRegex = `^[\w +\-=\.:/@]+$`
+	// GatewayEndpointRegex matches one or more word characters, optionally followed by dot-separated word segments
+	GatewayEndpointRegex = `^\w+(\.\w+)*$`
+
+	validateK8sResourceName        = combineValidationFuncs(regex(k8sResourceNameRegex), minLength(1), maxLength(253))
+	validateVpcID                  = combineValidationFuncs(regex(VpcIDRegex), maxLength(255))
+	validateEipAllocationID        = combineValidationFuncs(regex(EipAllocationIDRegex), maxLength(255))
+	validateSnapshotID             = combineValidationFuncs(regex(SnapshotIDRegex), maxLength(255))
+	validateIamInstanceProfileName = combineValidationFuncs(regex(IamInstanceProfileNameRegex), minLength(1), maxLength(128))
+	validateIamInstanceProfileArn  = combineValidationFuncs(regex(IamInstanceProfileArnRegex), maxLength(255))
+	validateZoneName               = combineValidationFuncs(regex(ZoneNameRegex), maxLength(255))
+	validateTagKey                 = combineValidationFuncs(regex(TagKeyRegex), minLength(1), maxLength(128))
+	validateGatewayEndpointName    = combineValidationFuncs(regex(GatewayEndpointRegex), maxLength(255))
+)
+
+type validateFunc[T any] func(T, *field.Path) field.ErrorList
+
+// combineValidationFuncs validates a value against a list of filters.
+func combineValidationFuncs[T any](filters ...validateFunc[T]) validateFunc[T] {
+	return func(t T, fld *field.Path) field.ErrorList {
+		var allErrs field.ErrorList
+		for _, f := range filters {
+			allErrs = append(allErrs, f(t, fld)...)
+		}
+		return allErrs
+	}
+}
+
+// regex returns a filterFunc that validates a string against a regular expression.
+func regex(regex string) validateFunc[string] {
+	compiled := regexp.MustCompile(regex)
+	return func(name string, fld *field.Path) field.ErrorList {
+		var allErrs field.ErrorList
+		if name == "" {
+			return allErrs // Allow empty strings to pass through
+		}
+		if !compiled.MatchString(name) {
+			allErrs = append(allErrs, field.Invalid(fld, name, fmt.Sprintf("does not match expected regex %s", compiled.String())))
+		}
+		return allErrs
+	}
+}
+
+// nolint:unparam
+func minLength(min int) validateFunc[string] {
+	return func(name string, fld *field.Path) field.ErrorList {
+		var allErrs field.ErrorList
+		if utf8.RuneCountInString(name) < min {
+			return field.ErrorList{field.Invalid(fld, name, fmt.Sprintf("must not be fewer than %d characters, got %d", min, len(name)))}
+		}
+		return allErrs
+	}
+}
+
+func maxLength(max int) validateFunc[string] {
+	return func(name string, fld *field.Path) field.ErrorList {
+		var allErrs field.ErrorList
+		if utf8.RuneCountInString(name) > max {
+			return field.ErrorList{field.Invalid(fld, name, fmt.Sprintf("must not be more than %d characters, got %d", max, len(name)))}
```

Comment:
Like in [this comment](https://github.com/gardener/gardener-extension-provider-aws/pull/1479/files#r2356342610) better use `utf8.RuneCountInString()` in the error message as well.